src/retry/validate.rs

```diff
@@ -0,0 +1,326 @@
+            chain_spec, shanghai, cancun, eip2718, eip1559, eip4844, block_gas_limit, max_tx_input_bytes, ..
+        } = self;
+
+        let inner = KakarotTransactionValidatorInner {
+            chain_spec,
+            client,
+            eip2718,
+            eip1559,
+            eip4844,
+            block_gas_limit,
+            max_tx_input_bytes,
+            _marker: Default::default(),
+        };
+
+        KakarotTransactionValidator { inner: Arc::new(inner) }
+    }
+}
+
+/// Validator for Ethereum transactions.
+#[derive(Debug, Clone)]
+pub struct KakarotTransactionValidator<P, T>
+where
+    P: EthereumProvider + Send + Sync,
+{
+    /// The type that performs the actual validation.
+    inner: Arc<KakarotTransactionValidatorInner<P, T>>,
+}
+
+impl<P, Tx> KakarotTransactionValidator<P, Tx>
+where
+    P: EthereumProvider + Send + Sync,
+{
+    /// Returns the configured chain spec
+    pub fn chain_spec(&self) -> Arc<ChainSpec> {
+        self.inner.chain_spec.clone()
+    }
+
+    /// Returns the configured client
+    pub fn client(&self) -> &EthDatabase<P> {
+        &self.inner.client
+    }
+}
+
+impl<P, Tx> KakarotTransactionValidator<P, Tx>
+where
+    P: EthereumProvider + Send + Sync,
+    Tx: EthPoolTransaction,
+{
+    /// Validates a single transaction.
+    ///
+    /// See also [`TransactionValidator::validate_transaction`]
+    pub fn validate_one(&self, transaction: Tx) -> TransactionValidationOutcome<Tx> {
+        self.inner.validate_one(transaction)
+    }
+
+    /// Validates all given transactions.
+    ///
+    /// Returns all outcomes for the given transactions in the same order.
+    ///
+    /// See also [`Self::validate_one`]
+    pub fn validate_all(&self, transactions: Vec<(TransactionOrigin, Tx)>) -> Vec<TransactionValidationOutcome<Tx>> {
+        transactions.into_iter().map(|(_origin, tx)| self.validate_one(tx)).collect()
+    }
+}
+
+impl<P, Tx> TransactionValidator for KakarotTransactionValidator<P, Tx>
+where
+    P: EthereumProvider + Send + Sync,
+    Tx: EthPoolTransaction,
+{
+    type Transaction = Tx;
+
+    async fn validate_transaction(
+        &self,
+        _origin: TransactionOrigin,
+        transaction: Self::Transaction,
+    ) -> TransactionValidationOutcome<Self::Transaction> {
+        self.validate_one(transaction)
+    }
+
+    async fn validate_transactions(
+        &self,
+        transactions: Vec<(TransactionOrigin, Self::Transaction)>,
+    ) -> Vec<TransactionValidationOutcome<Self::Transaction>> {
+        self.validate_all(transactions)
+    }
+
+    fn on_new_head_block(&self, _new_tip_block: &SealedBlock) {}
+}
+
+/// A [`TransactionValidator`] implementation that validates ethereum transaction.
+#[derive(Debug)]
+pub(crate) struct KakarotTransactionValidatorInner<P, T>
+where
+    P: EthereumProvider + Send + Sync,
+{
+    /// Spec of the chain
+    chain_spec: Arc<ChainSpec>,
+    /// This type fetches account info from the db
+    client: EthDatabase<P>,
```

Comment:
I wonder if we should use a `CachedDB<EthDatabase<P>>` here. The issue is that the since of the memory might grow if we don't ever prune accounts

Reply:
Yes could be a good idea, maybe in this "initial pr" we can keep it like that and open another one to clean this up, wdyt?

Reply:
fully aligned there; but in the validation of the mempool tx we must have access to the account to have its balance and check it. And unfortunately the only place where we do that is the provider...

The best strategy would surely be to put the mempool in the account manager service but we would have to find a method so that the send raw transaction has access to it too.